algorithms/HPdutility/R/hpdsample.R

```diff
@@ -0,0 +1,303 @@
+               else if (is.dframe(data)) dframe(dim = ddim, blocks = dblocks)
+
+  # We need metadata about which of the columns are factors so that we can
+  # keep track of factors locally within each partition.  If data is a dframe,
+  # dLevels is an array of the global levels for each column of data that is a
+  # factor. Otherwise, it is NULL and isn't used later
+  dLevels <- if(is.dframe(data)) .getGlobalLevels(data)
+
+  for (k in 1:npartitions(data)) { # For each partition dk of the input data
+    # Iterate over the output partitions si in parallel and copy some portion of
+    # the data from partition dk to sample partition si. This copied data will
+    # be appended to the data copied from previous partitions
+    foreach(i, 1:nSamplePartitions, function(si = splits(sdata, i),
+                                      dk = splits(data, k),
+                                      psizes = psizes,
+                                      dLevels = dLevels,
+                                      k = k, 
+                                      .coerceFactorsToGlobalLevels =
+                                        .coerceFactorsToGlobalLevels, 
+                                      .copyClassStructure = .copyClassStructure) {
+      # Create a random index to choose which data items to copy from dk to si
+      idx <- sample(1:nrow(dk), psizes[k], replace = TRUE)
+
+      # Based on the value of k we decide where to copy the data in splits si
+      # since we don't want to overwrite data that was previously copied.
+      # Effectively appends this data to previous data
+      startRow <- if (k == 1) 1 else sum(psizes[1:(k - 1)]) + 1 
+      endRow   <- startRow + length(idx) - 1
+
+      if (is.data.frame(dk)) {
+        # If dLevels is non-null, need to convert its factor columns to use the
+        # same levels map as the global levels in case the levels aren't
+        # consistent from partition to partition of 'data'
+        if (!is.null(dLevels)) 
+          dk <- .coerceFactorsToGlobalLevels(dk, dLevels)
+
+        # First iteration, no data has been copied yet. Ensure all the types of
+        # the sample data are the same as the originals
+        if (k == 1) 
+          si <- .copyClassStructure(dk, si, dLevels)
+      }
+
+      # Copy data items into sampled data
+      si[startRow:endRow,] <- dk[idx,,drop = F]
+
+      update(si)
+    }, progress = trace, scheduler = 1)
+  }
+  colnames(sdata) <- colnames(data)
+  sdata 
+}
+
+.hpdsampledual <- function(data1, data2, nSamplePartitions, samplingRatio, trace) {
+  # Function to sample two darrays/dframes in conjunction
+  
+  if (!(is.darray(data1) || is.dframe(data1))) 
+    stop("'data1' must be a darray or dframe")
+
+  if (!(is.darray(data2) || is.dframe(data2))) 
+    stop("'data2' must be a darray or dframe")
+
+  if (nrow(data1) != nrow(data2)) 
+    stop("'data1' and 'data2' must be the same length")
+
+  if (npartitions(data1) != npartitions(data2)) 
```

Comment:
Instead of this check and the previous one, please add two checks that (1) data1 is only row-wise partitioned, (2) data1 and data2 have the same partition sizes.
You may notice that having the same nrow and npartitions does not guarantee that the two dobject have the same partitioning schema.